src/fuzz.rs

```diff
@@ -0,0 +1,98 @@
+use bls_crypto::hashers::DirectHasher;
+use rand_chacha::ChaCha8Rng;
+use rand_core::SeedableRng;
+use threshold_bls::curve::bls12377::PairingCurve as bls377;
+use threshold_bls::group::Element;
+
+use crate::api::{Poly, PoprfScheme, PrfScheme, Scheme, Share, ThresholdScheme};
+use crate::bls12_377::Scalar;
+use crate::hash_to_field::{HashToField, TryAndIncrement};
+//use crate::PoprfError;
+
+type G2Scheme = crate::poprf::G2Scheme<bls377>;
+
+pub fn hash_to_field(data: &[u8]) {
+    let domain = b"H2FFUZZ";
+    let hasher = TryAndIncrement::<_, Scalar>::new(&DirectHasher);
+    let hash = hasher.hash_to_field(domain, data).unwrap();
+    assert!(hash != Scalar::one());
+    assert!(hash != Scalar::zero());
+}
+
+// This struct exists to make it easy for the fuzzer to know what sort of
+// inputs to provide. We allow the fuzzer to seed the RNG, which is
+// deterministic anyway (we only seed it to avoid code changes),
+// and supply the message and tag.
+#[derive(arbitrary::Arbitrary, Debug)]
+pub struct PoprfFuzzInput {
+    msg: Vec<u8>,
+    tag: Vec<u8>,
+    seed: [u8; 32],
+    terms: u8,
+}
+
+pub fn poprfscheme(input: PoprfFuzzInput) {
+    //let mut rng = rand::thread_rng();
+    let mut rng = ChaCha8Rng::from_seed(input.seed);
+
+    let t: usize = (input.terms % 27) as usize + 3;
```

Comment:
Where does this equation come from?

Reply:
I can answer this as I wrote the original code.

Essentially I wanted to exercise the number of participants "reasonably" well, to see if we allowed the fuzzer to vary (t,n), if that triggers anything interesting. However, the fuzzer left to its own devices will pick '0' quite quickly and start reporting crashes, while 0 is in fact just an invalid use of the API.

An alternative is to modify the code to accept 0 but rather than panic, return failure. If we could catch panics in Rust, we could also catch and suppress it here, but not continue the code. The fuzzer would exercise this code path, but not report a crash (any panic is equivalent to a crash for the fuzzer because it triggers and unwind).

So I picked a range out of thin air, really: 3-30 terms, see how the code deals with it by allowing the fuzzer to vary it.

You could actually do the same thing in unit testing: make your existing unit test a helper function that accepts arbitrary usize, and create a new test function that calls the helper over some range, to check that this works.